### ISSUE TITLE
Exchange Token for clients within same Realm throws access_denied

### ISSUE THREAD

**Issue by happy0088** [keycloak/keycloak#17008]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

token-exchange

### Describe the bug

I am upgrading from V14 to V20 .

logged in user - [abc@xyz.com](mailto:abc@xyz.com)
generate a token as a public client using client1 .
Try to run an exchange token call with client1 - returns proper token with
```
"aud": [
"account",
"client1"
]
```
When i try to run an exchange token with another client2 - it throws error
`error=not_allowed, reason='client is not the token holder', auth_method=token_exchange, grant_type=urn:ietf:params:oauth:grant-type:token-exchange, client_auth_method=client-secret",`
There is one observation that the public client token carries client1 entry as
"azp": "client1" (since the token was generated using client1)

i have the features enabled :
```

extraEnvVars:
  - name: KC_FEATURES
    value: admin-fine-grained-authz,token-exchange,scripts
```

This functionality was working perfectly fine in V14 but started throwing error with V20.

I came across this below documentation from keycloak which talks about
i) explicit permission setup for the target client(client2 i believe in our case ) (not doing this for V14) and
ii) sending extra params in the header for the exchange calls . (which we are already doing even in V14)
https://www.keycloak.org/docs/latest/securing_apps/index.html#internal-token-to-internal-token-exchange
(tried following the steps for a client via admin console but couldn't get it to work)

This seems little difficult in my scenario since we are creating clients on the fly from application and doing the permission registration will be very difficult

### Version

20.0.3

### Expected behavior

should generate the exchange token

### Actual behavior

Error

### How to Reproduce?

enable the features and try to generate the exchange token with a different client not present in the bearer token for the request .

**Comment by bpedersen2** [comment]
Check https://www.keycloak.org/docs/latest/securing_apps/index.html#expand-permission-model-with-service-accounts

**Comment by happy0088** [comment]
Thanks @bpedersen2 . 
The above link talks about an alternative approach to achieve this . ** tried the approach  ,but  i am unable to even get this working via explicit admin console settings under the permission tab (i can share the details as needed for the settings ).**
However  , the same is working without any permission-policy setup in V14 . This makes me curious to understand if there is any explicit mandate that has been forced post V14  .

**Comment by happy0088** [comment]
Hi @bpedersen2  .. checking if there can be any more help on the above issue ?

**Comment by mposolda** [comment]
Thanks for the report, but unfortunately due the amount of other reported issues and other priorities, Keycloak team does not have time to properly triage this bug. So preliminary added to Backlog for now.
It will be helpful if:
- You can verify if still applicable in latest Keycloak released version. If not, then it is welcome to close this issue.
- If you figure that this may not be a valid bug (for example just a mistake in configuration etc), it will be also welcome to close this issue
- If still applicable in latest version, it will be welcome to add the comment as well, that this was still reproduced with latest Keycloak version as it is very valuable info. Anyone is welcome to comment with this or add other relevant comments to this issue.

**Comment by bclouser** [comment]
I was able to reproduce this in version 21.1.1. I have the preview features `token-exchange` and `admin-fine-grained-authz` enabled, and i see the error as described by @happy0088.

I am still angling for a workaround, but currently it does look like this feature broke

**Comment by ActJV** [comment]
Facing the same issue on version 20.0.3 with preview features `token-exchange` and `admin-fine-grained-authz` enabled.

@bclouser any update on a workaround?

**Comment by dteleguin** [comment]
The issue has been clearly introduced by #11542. It looks like the relationships between the clients participating in token exchange are becoming even more complex. We have issuing client (aka token holder), exchanging client and target client (aka audience); the rules are different for public and confidential clients, and everything is hardcoded into the [DefaultTokenExchangeProvider](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/DefaultTokenExchangeProvider.java#L317).

At the same time, the following looks like a valid use case:
* clients A and C are public;
* client B is confidential and has special permissions;
* client B should be able to exchange tokens issued by A, set audience to C (`audience=...`) and also perform impersonation (`requested_subject=...`).

Could there potentially be a better solution than just hardcoding the rules? Could we leverage fine grained admin permissions here? WDYT @pedroigor ?

**Comment by dteleguin** [comment]
As for workaround, the only viable option at the moment seems to be deploying a custom `TokenExchangeProvider` that would override `exchangeClientToClient()` and bring in the old (pre-18) behavior. @ActJV @Mica4DEV @judielsb

**Comment by pedroigor** [comment]
Hey @dteleguin. Why would you want to exchange a token from a confidential client where the audience is a public client?

**Comment by FloHin** [comment]
For the last couple of days, I also investigated many approaches for a similar setup, and many different approaches exist, but seem to be outdated with current versions of keycloak, some mention after V14 it was not working any more

The problem for us is the refreshing. e.g when using a legacy system, with a necessary backend login.
* use a confidential client A to authenticate directly at keycloak.
* exchange token to obtain a refresh token for client B, so that refreshing afterwards *would* be possible
* deliver the exchanged token for B to the user's app

Errors are usually of the following:
* "{"error":"access_denied","error_description":"Client is not the holder of the token"}"
* "{"error":"invalid_grant","error_description":"Invalid refresh token. Token client and authorized client don't match"}"

I am sorry if this is not helpful for debugging, just wanted to point out that there seem to be many people struggling with that without proper solutions.

The **azp** is always the confidential client A, unusable by client B

**Comment by pedroigor** [comment]
@FloHin Yeah, I see quite a few people are stuck because of the same problem. But still, I'm not sure about the use cases we are trying to solve here. I just want to avoid to introduce something that can be exploited as we have in the past. We can't put much trust in public clients.

If you can provide request samples to understand why those errors are happening, I appreciate.

**Comment by ActJV** [comment]
Enabled features: `fine-grain-authz` + `token-exchange`

2 Clients are configured:
- my-client
- my-confidential-client

an audience-mapper is applied on the my-confidential-client to the my-client.
- `impersonate` and `user-impersonated` policies are applied to the my-client.
- `token-exchange` policy is applied to the my-confidential-client.

Here is a request sample in PowerShell:
```
# Configure your client credentials
$clientId = "my-confidential-client"
$clientSecret = "my-confidential-client-secret"
$baseEndpoint = "https://keycloak.domain.org/auth/realms/my-reals"
$tokenEndpoint = $baseEndpoint + "/protocol/openid-connect/token"
$finalClientId = "my-client"
$userSubject = "subject_userName"

# Step 1: Request an OAuth 2.0 token using confidential client credentials
$tokenRequestBody = @{
    grant_type = "client_credentials"
    client_id = $clientId
    client_secret = $clientSecret
}

$tokenResponse = Invoke-RestMethod -Uri $tokenEndpoint -Method Post -ContentType "application/x-www-form-urlencoded" -Body $tokenRequestBody

# Extract the access token for the confidential-client from the response
$accessToken = $tokenResponse.access_token

# Step 2: Request a token exchange using the obtained access token
$tokenExchangeRequestBody = @{
    grant_type = "urn:ietf:params:oauth:grant-type:token-exchange"
	client_id = $finalClientId
    requested_subject = $userSubject
    subject_token = $accessToken
}

$tokenExchangeResponse = Invoke-RestMethod -Uri $tokenEndpoint -Method Post -ContentType "application/x-www-form-urlencoded" -Body $tokenExchangeRequestBody

# Extract the exchanged token from the response
$exchangedToken = $tokenExchangeResponse.access_token

Write-Host "Exchanged token: $exchangedToken"
```

**Comment by raubel** [comment]
It looks like I have exactly the same issue with Keycloak 23.0.5 and those features: declarative-user-profile, token-exchange, admin-fine-grained-authz.

I'd like to explain my use case because maybe I'm wrong and I should review my solution.

I have a service S exposing a REST API protected by a public (non confidential) client. A user can use this REST API by using the 'password' grant with this client, providing their own username and password. This client is also in charge of populating the access tokens with various data (roles, groups, profile) and is then configured with some specific mappers.

I also have a UI and its BFF having its own confidential client with the Standard Authentication Flow set to ON, so that the user can authenticate to the UI using the Authorization Code Flow.
This API needs to get information from S on behalf of the user. As I want the token validated by S to have the expected format (roles, groups, profile), my idea is for the UI confidential client to exchange its token with a token delivered by the client of S.

Did I miss something?

**Comment by duboiol** [comment]
I observed a similar issue with both 25.0.0 and 26.0.7, 

The token chain is:

1. Client-A exchanges an external id_token to oauth2 access token, audience client-B
2. Client-B exchanges the access token he is in the audience.

 AuthenticationManager.verifyIdentityToken( in tokenExchange()) fails (code 400, "invalid token", event s"ubject_token validation failure").

What happens is that in 
 client = realm.getClientByClientId(token.getIssuedFor()); fails, as the token verification  because the audience client is not the client that generated the token.
 
 The flag isCookie to verifyIdentityToken disable the client identity check. As the token exchange performs later the check on the issuer and the audience, this is safe (at least in this internal to internal case).
 
 With isCookie to true, token exchange works. 
 
 There is little info on the semantic of the flag but to me this is the only side effect, bypassing the client verification.

**Comment by mposolda** [comment]
Closing as token-exchange was updated for Keycloak 26.2.0 and is officially supported now. The implementation is rewritten and the message `client is not the token holder` is not thrown anymore by the standard token exchange. Please open follow-up issue if you want to support some use-case, which is not supported by Keycloak 26.2.0 standard token exchange.
